File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
void solve(vector<int>& temp,vector<vector<int>>& results,vector<int>& candidates,int index,int target,int n) {
    if(target==0) {
        results.push_back(temp);
        return;
    }
    for(int i=index;i<n;i++) {
         if(i!=index && candidates[i]==candidates[i-1]) {
                //Remove the duplicate combination case
                continue;
            }
        if(target-candidates[i]>=0) {
           
            temp.push_back(candidates[i]);
            solve(temp,results,candidates,i+1,target-candidates[i],n);
            temp.pop_back();
        }
    }
    
    
}
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> results;
        vector<int> temp;
        sort(candidates.begin(),candidates.end());
        solve(temp,results,candidates,0,target,candidates.size());
        return results;

    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        answer = []
        curr_answer = []
        candidates.sort()
        def solution(index,new_target,curr_answer):
            if new_target<0: return
            if new_target==0:
                answer.append(curr_answer.copy())
                return
            #pick or no pick
            for i,num in enumerate(candidates[index:]):
                #pick
                if i>0 and candidates[i+index]==candidates[i-1+index]: continue
                curr_answer.append(num)
                solution(index+i+1,new_target-num,curr_answer)
                curr_answer.pop()
        solution(0,target,curr_answer)
        return answer


            